services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/ImmutableTableSchema.java

```diff
@@ -0,0 +1,327 @@
+import software.amazon.awssdk.enhanced.dynamodb.TableSchema;
+import software.amazon.awssdk.enhanced.dynamodb.internal.immutable.ImmutableInfo;
+import software.amazon.awssdk.enhanced.dynamodb.internal.immutable.ImmutableIntrospector;
+import software.amazon.awssdk.enhanced.dynamodb.internal.immutable.ImmutablePropertyDescriptor;
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.BeanAttributeGetter;
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.BeanAttributeSetter;
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.ObjectConstructor;
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.ObjectGetterMethod;
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.StaticGetterMethod;
+import software.amazon.awssdk.enhanced.dynamodb.internal.mapper.WrappedTableSchema;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.BeanTableSchemaAttributeTag;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbAttribute;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbBean;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbConvertedBy;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbFlatten;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbImmutable;
+
+/**
+ * Implementation of {@link TableSchema} that builds a table schema based on properties and annotations of an immutable
+ * class with an associated builder class. Example:
+ * <pre>
+ * <code>
+ * {@literal @}DynamoDbImmutable(builder = Customer.Builder.class)
+ * public class Customer {
+ *     {@literal @}DynamoDbPartitionKey
+ *     public String accountId() { ... }
+ *
+ *     {@literal @}DynamoDbSortKey
+ *     public int subId() { ... }
+ *
+ *     // Defines a GSI (customers_by_name) with a partition key of 'name'
+ *     {@literal @}DynamoDbSecondaryPartitionKey(indexNames = "customers_by_name")
+ *     public String name() { ... }
+ *
+ *     // Defines an LSI (customers_by_date) with a sort key of 'createdDate' and also declares the
+ *     // same attribute as a sort key for the GSI named 'customers_by_name'
+ *     {@literal @}DynamoDbSecondarySortKey(indexNames = {"customers_by_date", "customers_by_name"})
+ *     public Instant createdDate() { ... }
+ *
+ *     // Not required to be an inner-class, but builders often are
+ *     public static final class Builder {
+ *         public Builder accountId(String accountId) { ... };
+ *         public Builder subId(int subId) { ... };
+ *         public Builder name(String name) { ... };
+ *         public Builder createdDate(Instant createdDate) { ... };
+ *
+ *         public Customer build() { ... };
+ *     }
+ * }
+ *
+ * </pre>
+ * @param <T> The type of object that this {@link TableSchema} maps to.
+ */
+@SdkPublicApi
+public final class ImmutableTableSchema<T> extends WrappedTableSchema<T, StaticImmutableTableSchema<T, ?>> {
```

Comment:
I am thinking I should add 'implements TableSchema<T>' here because WrappedTableSchema is an internal class and I want the TableSchema interface to be part of the public contract of this class.